Core/HLE/sceNet.cpp

```diff
@@ -191,11 +233,26 @@ void __NetDoState(PointerWrap &p) {
 		p.Do(netThread1Addr);
 		p.Do(netThread2Addr);
 	}
+	if (s >= 4) {
+		p.Do(netApctlState);
+		p.Do(netApctlInfo);
+		p.Do(actionAfterApctlMipsCall);
+		__KernelRestoreActionType(actionAfterApctlMipsCall, AfterApctlMipsCall::Create);
+		p.Do(apctlThreadHackAddr);
+	}
```

Comment:
We should probably put Apctl related vars in "sceNetApctl" section instead of "sceNet" section like this, but then again `netApctlInited` and `apctlHandlers` have been part of "sceNet" section for a long time.

Reply:
If s is < 4 you still need to do something about actionAfterApctlMipsCall.  Otherwise bad things can happen when people load older states.  This might be why you got the load state corruption in your other comment.

Generally should always have an else setting default values or something on these `s >= N` checks.

-[Unknown]

Reply:
Well if  s < 4 i want to leave it as default (the one came from `__NetInit` which should already have a valid value), 
or should i re-init those value? ie. Re-register the action type `actionAfterApctlMipsCall = __KernelRegisterActionType(AfterApctlMipsCall::Create);` within DoState (does registering the same PSPAction like this allowed?)

Reply:
No, the value from __NetInit is no longer valid because we've loaded a save state.  All these values become invalid now.  That's why __KernelRestoreActionType exists.

In this case, you'd end up with something like...

```c++
... DoState(PointerWrap &p) {
    if (s >= 123) {
        p.Do(actionAfterFoo); // Save/load the action ID used per this state.
        __KernelRestoreActionType(actionAfterFoo, AfterFoo::Create); // Reassign the AfterFoo action to that ID.
    } else {
        // That state didn't have Foo, so we don't have an ID.
        // We cannot allocate one now, as we'd conflict with someone else's.
        actionAfterFoo = 0;
    }
}

...

// On use.
if (!actionAfterFoo) {
    // We must've come from a save state.  Let's allocate now.
    actionAfterFoo = __KernelRegisterActionType(AfterFoo::Create);
}
... = __KernelCreateAction(actionAfterFoo);
```

I realize it's a bit of a pain, but save states are a bit complicated.

-[Unknown]

Reply:
Hmm.. i don't see something like `if (!actionAfterxxx)` on sceFont and sceMpeg before they use it on `__KernelCreateAction` they only register it once during `__FontInit` and `__MpegInit` similar to mine that got registered during `__NetInit`